### ISSUE TITLE
FR: A way to follow links in documents on non-touch devices

### ISSUE THREAD

**Issue by vovavili** [koreader/koreader#6562]
**Does your feature request involve difficulty completing a task? Please describe.**
My Kindle 4 NT has neither has a physical keyboard nor a touchscreen, which makes it impossible to follow links inside documents in any format in KOReader.

**Describe the solution you'd like**
Some shortcut which enables you to cycle through links on a page you're currently viewing - for example, in [CoolReader 3 for Kindle](https://github.com/CrazyCoder/coolreader-kindle-qt), all you had to do was press Keyboard+Up.

**Comment by poire-z** [comment]
I don't know anything about how keys/pad are used on such non-touch devices, and what's still available or what would conflict or not, and how to trigger or go back.
But if one wants to go on with this, here's some patch to readerlink to allow cycling thru page links, and to jump to the selected one - on CreDocuments only (EPUB, FB2):
```diff
--- a/frontend/apps/reader/modules/readerlink.lua
+++ b/frontend/apps/reader/modules/readerlink.lua
@@ -24,6 +24,18 @@ local ReaderLink = InputContainer:new{
 }

 function ReaderLink:init()
+    if Device:hasKeys() then
+        self.key_events.SelectNextPageLink = {
+            { {"L" } },
+            doc = "select next page link",
+            event = "SelectNextPageLink",
+        }
+        self.key_events.GotoSelectedPageLink = {
+            { { "G" } },
+            doc = "go to selected page link",
+            event = "GotoSelectedPageLink",
+        }
+    end
     if Device:isTouchDevice() then
         self.ui:registerTouchZones({
             {
@@ -937,6 +949,75 @@ function ReaderLink:onGoToInternalPageLink(ges)
     self:onGoToPageLink(ges, true)
 end

+function ReaderLink:onSelectNextPageLink()
+    if self.ui.document.info.has_pages then
+        return
+    end
+    local links = self.ui.document:getPageLinks()
+    if not links or #links == 0 then
+        return
+    end
+    if self.cur_selected_page_link_num then
+        self.cur_selected_page_link_num = self.cur_selected_page_link_num + 1
+        if self.cur_selected_page_link_num > #links then
+            self.cur_selected_page_link_num = 1
+        end
+    else
+        self.cur_selected_page_link_num = 1
+    end
+    local selected_link = links[self.cur_selected_page_link_num]
+    logger.dbg("selected_link", selected_link)
+    -- Check a_xpointer is coherent, use it as from_xpointer only if it is
+    local from_xpointer = nil
+    if selected_link.a_xpointer and self:isXpointerCoherent(selected_link.a_xpointer) then
+        from_xpointer = selected_link.a_xpointer
+    end
+    local link_y
+    if selected_link.segments and #selected_link.segments > 0 then
+        link_y = selected_link.segments[#selected_link.segments].y1
+    else
+        link_y = selected_link.end_y
+    end
+    -- Make it a link as expected by onGotoLink
+    self.cur_selected_link = {
+        xpointer = selected_link.section or selected_link.uri,
+        marker_xpointer = selected_link.section,
+        from_xpointer = from_xpointer,
+        -- (keep a_xpointer even if incoherent, might be needed for
+        -- footnote detection (better than nothing if incoherent)
+        a_xpointer = selected_link.a_xpointer,
+        -- keep the link y position, so we can keep its highlight shown
+        -- a bit more time if it was hidden by the footnote popup
+        link_y = link_y,
+    }
+    self.ui.document:highlightXPointer() -- clear any previous one
+    self.ui.document:highlightXPointer(self.cur_selected_link.from_xpointer)
+    UIManager:setDirty(self.dialog, "ui")
+    return true
+end
+
+function ReaderLink:onGotoSelectedPageLink()
+    if self.cur_selected_link then
+        return self:onGotoLink(self.cur_selected_link, false, isFootnoteLinkInPopupEnabled())
+    end
+end
+
+function ReaderLink:onPageUpdate()
+    if self.cur_selected_link then
+        self.ui.document:highlightXPointer()
+        self.cur_selected_page_link_num = nil
+        self.cur_selected_link = nil
+    end
+end
+
+function ReaderLink:onPosUpdate()
+    if self.cur_selected_link then
+        self.ui.document:highlightXPointer()
+        self.cur_selected_page_link_num = nil
+        self.cur_selected_link = nil
+    end
+end
+
 function ReaderLink:onGoToLatestBookmark(ges)
     local latest_bookmark = self.ui.bookmark:getLatestBookmark()
     if latest_bookmark then
```

**Comment by Frenzie** [comment]
@poire-z Trigger would be with `Press` (mapped to Enter on the emulator). Back with… back (Escape on the emulator). ;-)

**Comment by vovavili** [comment]
@poire-z this is so nice, thanks!

**Comment by poire-z** [comment]
Probably, dunno, I'm not familiar with the hasKeyPad/DPAD/hasFewKeys stuff - and I don't have such a device.
Looking in some other modules, I see up/left might both be used for going to previous page and down/right for going to next page. Is that the case?
If yes, well, I dunno how much people are different and using up/down or left/right, and changing one of these for some other function would affect these people.

**Comment by poire-z** [comment]
@vovavili : if you want to test this quick modification of #6619 , just replace
`koreader/frontend/apps/reader/modules/readerlink.lua` with [readerlink.lua.txt](https://github.com/koreader/koreader/files/5158971/readerlink.lua.txt) (remove the .txt extension).
Quick and dirty mods:
```diff
 function ReaderLink:init()
-    if Device:hasKeys() then
+    if Device:hasDPad() then
         self.key_events.SelectNextPageLink = {
-            {"Tab" },
+            {"Down" },
             doc = "select next page link",
             event = "SelectNextPageLink",
         }
         self.key_events.SelectPrevPageLink = {
-            {"Shift", "Tab" },
-            {"Sym", "Tab" }, -- Right Shift + Tab
+            {"Up" },
             doc = "select previous page link",
             event = "SelectPrevPageLink",
         }
```

**Comment by vovavili** [comment]
I can confirm that this quick-and-dirty method works as intended on my K4NT, nice work my dude.

**Comment by poire-z** [comment]
Solved by #8877.

**Comment by Commodore64user** [comment]
properly adding support for this with #11749 on kindle. `ScreenKB` + `LPgBack`/`LPgFwd` for k4, `Shift` + `LPgBack`/`LPgFwd` for the rest.
